File: lib/models/employee.dart
/// Employment type: part-time, contract, or student
enum EmploymentType { partTime, contract, permanent, student }
enum EmployeeAccessLevel { employee, manager, hrAdmin }

/// Whether the worker is on staff or third-party. Backed by
/// `employees.worker_type` (added by the worker_type_and_invites
/// migration). Drives permissions and reporting (contractor scorecard).
enum WorkerType { employee, contractor, subcontractor }

extension WorkerTypeX on WorkerType {
  String get wireValue {
    switch (this) {
      case WorkerType.employee:
        return 'employee';
      case WorkerType.contractor:
        return 'contractor';
      case WorkerType.subcontractor:
        return 'subcontractor';
    }
  }

  String get label {
    switch (this) {
      case WorkerType.employee:
        return 'Employee';
      case WorkerType.contractor:
        return 'Contractor';
      case WorkerType.subcontractor:
        return 'Subcontractor';
    }
  }
}

WorkerType workerTypeFromString(String? raw) {
  switch ((raw ?? '').toLowerCase()) {
    case 'contractor':
      return WorkerType.contractor;
    case 'subcontractor':
      return WorkerType.subcontractor;
    default:
      return WorkerType.employee;
  }
}

class Employee {
  final String name;
  final String surname;
  final String id;
  final String employeeCode;
  final DateTime employmentDate;
  final EmploymentType employmentType;
  final String position;
  final double monthlySalary;
  final double hourlyRate;
  final double weeklyRate;
  final double dailyRate;
  final double overtimeRate;
  final double doubleTimeRate;
  final double workDaysWeekly;
  final double dailyHours;
  final String branch;
  final String? managerUserId;
  final EmployeeAccessLevel accessLevel;
  final String? employmentTypeLabel;
  final String? email;
  final String? phone;
  final String? profileId;
  final WorkerType workerType;

  const Employee({
    required this.name,
    required this.surname,
    required this.id,
    this.employeeCode = '',
    required this.employmentDate,
    required this.employmentType,
    required this.position,
    this.monthlySalary = 0,
    this.hourlyRate = 0,
    this.weeklyRate = 0,
    this.dailyRate = 0,
    this.overtimeRate = 0,
    this.doubleTimeRate = 0,
    this.workDaysWeekly = 5,
    this.dailyHours = 8,
    this.branch = '',
    this.managerUserId,
    this.accessLevel = EmployeeAccessLevel.employee,
    this.employmentTypeLabel,
    this.email,
    this.phone,
    this.profileId,
    this.workerType = WorkerType.employee,
  });

  String get fullName => '$name $surname'.trim();
  String get displayName => fullName.isNotEmpty ? fullName : (employeeCode.isNotEmpty ? employeeCode : id);

  static const int standardHoursPerDay = 8;

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'surname': surname,
      'id': id,
      'employee_code': employeeCode,
      'employment_date': employmentDate.toIso8601String(),
      'employment_type': switch (employmentType) {
        EmploymentType.contract => 'contract',
        EmploymentType.student => 'student',
        _ => 'part-time',
      },
      'position': position,
      'monthly_salary': monthlySalary,
      'hourly_rate': hourlyRate,
      'weekly_rate': weeklyRate,
      'daily_rate': dailyRate,
      'Overtime_rate': overtimeRate,
      'double_time_rate': doubleTimeRate,
      'work_days_weekly': workDaysWeekly,
      'daily_hours': dailyHours,
      'branch': branch,
      'manager_user_id': managerUserId,
      'access_level': switch (accessLevel) {
        EmployeeAccessLevel.manager => 'manager',
        EmployeeAccessLevel.hrAdmin => 'hr_admin',
        _ => 'employee',
      },
      'employment_type_label': employmentTypeLabel,
      if (email != null && email!.isNotEmpty) 'email': email!.trim().toLowerCase(),
      if (phone != null && phone!.isNotEmpty) 'phone': phone!.trim(),
    };
  }

  factory Employee.fromMap(Map<String, dynamic> map) {
    final typeStr = (map['employment_type'] as String? ?? '').toLowerCase();
    final type = typeStr.contains('contract')
        ? EmploymentType.contract
        : typeStr.contains('permanent')
            ? EmploymentType.permanent
        : typeStr.contains('student')
            ? EmploymentType.student
            : EmploymentType.partTime;
    return Employee(
      name: map['name'] as String? ?? '',
      surname: map['surname'] as String? ?? '',
      id: map['id']?.toString() ?? '',
      employeeCode: map['employee_code'] as String? ?? '',
      employmentDate: map['employment_date'] != null
          ? DateTime.tryParse(map['employment_date'] as String) ?? DateTime.now()
          : DateTime.now(),
      employmentType: type,
      position: map['position'] as String? ?? '',
      monthlySalary: (map['monthly_salary'] as num?)?.toDouble() ?? (map['monthly'] as num?)?.toDouble() ?? 0,
      hourlyRate: (map['hourly_rate'] as num?)?.toDouble() ?? (map['hourly'] as num?)?.toDouble() ?? 0,
      weeklyRate: (map['weekly_rate'] as num?)?.toDouble() ?? 0,
      dailyRate: (map['daily_rate'] as num?)?.toDouble() ?? 0,
      overtimeRate: (map['Overtime_rate'] as num?)?.toDouble() ?? 0,
      doubleTimeRate: (map['double_time_rate'] as num?)?.toDouble() ?? 0,
      workDaysWeekly: (map['work_days_weekly'] as num?)?.toDouble() ?? 5,
      dailyHours: (map['daily_hours'] as num?)?.toDouble() ?? 8,
      branch: map['branch'] as String? ?? '',
      managerUserId: map['manager_user_id']?.toString(),
      accessLevel: switch ((map['access_level'] as String? ?? 'employee').toLowerCase()) {
        'manager' => EmployeeAccessLevel.manager,
        'hr_admin' => EmployeeAccessLevel.hrAdmin,
        _ => EmployeeAccessLevel.employee,
      },
      employmentTypeLabel: map['employment_type_label'] as String?,
      email: map['email'] as String?,
      phone: map['phone'] as String?,
      profileId: map['profile_id']?.toString(),
    );
  }

  Employee copyWith({
    String? name,
    String? surname,
    String? id,
    String? employeeCode,
    DateTime? employmentDate,
    EmploymentType? employmentType,
    String? position,
    double? monthlySalary,
    double? hourlyRate,
    double? weeklyRate,
    double? dailyRate,
    double? overtimeRate,
    double? doubleTimeRate,
    double? workDaysWeekly,
    double? dailyHours,
    String? branch,
    String? managerUserId,
    EmployeeAccessLevel? accessLevel,
    String? employmentTypeLabel,
    String? email,
    String? phone,
    String? profileId,
    WorkerType? workerType,
  }) {
    return Employee(
      name: name ?? this.name,
      surname: surname ?? this.surname,
      id: id ?? this.id,
      employeeCode: employeeCode ?? this.employeeCode,
      employmentDate: employmentDate ?? this.employmentDate,
      employmentType: employmentType ?? this.employmentType,
      position: position ?? this.position,
      monthlySalary: monthlySalary ?? this.monthlySalary,
      hourlyRate: hourlyRate ?? this.hourlyRate,
      weeklyRate: weeklyRate ?? this.weeklyRate,
      dailyRate: dailyRate ?? this.dailyRate,
      overtimeRate: overtimeRate ?? this.overtimeRate,
      doubleTimeRate: doubleTimeRate ?? this.doubleTimeRate,
      workDaysWeekly: workDaysWeekly ?? this.workDaysWeekly,
      dailyHours: dailyHours ?? this.dailyHours,
      branch: branch ?? this.branch,
      managerUserId: managerUserId ?? this.managerUserId,
      accessLevel: accessLevel ?? this.accessLevel,
      employmentTypeLabel: employmentTypeLabel ?? this.employmentTypeLabel,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      profileId: profileId ?? this.profileId,
      workerType: workerType ?? this.workerType,
    );
  }
}

File: KaiFlow.Timesheets.Maui/Models/Employee.cs
using KaiFlow.Timesheets.Helpers;
using Newtonsoft.Json;
using Supabase.Postgrest.Attributes;
using Supabase.Postgrest.Models;

namespace KaiFlow.Timesheets.Models;

public enum EmploymentType { PartTime, Contract, Permanent, Student }
public enum AccessLevel { Employee, Manager, Admin, HrAdmin, Owner }
public enum WorkerType { Employee, Contractor, Subcontractor }

[Table("employees")]
public class Employee : BaseModel
{
    [PrimaryKey("id")]
    public Guid Id { get; set; }

    [Column("name")]
    public string Name { get; set; } = "";

    [Column("surname")]
    public string Surname { get; set; } = "";

    [Column("employee_code")]
    public string? EmployeeCode { get; set; }

    [Column("employment_type")]
    public string EmploymentTypeRaw { get; set; } = "permanent";

    [Column("access_level")]
    public string AccessLevelRaw { get; set; } = "employee";

    [Column("worker_type")]
    public string WorkerTypeRaw { get; set; } = "employee";

    [Column("position")]
    public string? Position { get; set; }

    [Column("branch")]
    public string? Branch { get; set; }

    [Column("employment_date")]
    public DateOnly? EmploymentDate { get; set; }

    [Column("hourly_rate")]
    public double HourlyRate { get; set; }

    [Column("daily_rate")]
    public double DailyRate { get; set; }

    [Column("weekly_rate")]
    public double WeeklyRate { get; set; }

    [Column("monthly_salary")]
    public double MonthlySalary { get; set; }

    [Column("overtime_rate")]
    public double OvertimeRate { get; set; }

    [Column("double_time_rate")]
    public double DoubleTimeRate { get; set; }

    [Column("pay_basis")]
    public string? PayBasisRaw { get; set; }

    [Column("paye_rate_percent")]
    public double? PayeRatePercent { get; set; }

    [Column("paye_fixed_amount")]
    public double PayeFixedAmount { get; set; }

    [Column("uif_exempt")]
    public bool UifExempt { get; set; }

    [Column("uif_rate_percent")]
    public double? UifRatePercent { get; set; }

    [Column("uif_fixed_amount")]
    public double UifFixedAmount { get; set; }

    [Column("tax_number")]
    public string? TaxNumber { get; set; }

    [Column("paye_reference")]
    public string? PayeReference { get; set; }

    [Column("medical_aid_member_number")]
    public string? MedicalAidMemberNumber { get; set; }

    [Column("pension_fund_number")]
    public string? PensionFundNumber { get; set; }

    [Column("tax_directive_number")]
    public string? TaxDirectiveNumber { get; set; }

    [Column("tax_directive_rate_percent")]
    public double? TaxDirectiveRatePercent { get; set; }

    [Column("date_of_birth")]
    public DateOnly? DateOfBirth { get; set; }

    [Column("cost_center")]
    public string? CostCenter { get; set; }

    [Column("termination_date")]
    public DateOnly? TerminationDate { get; set; }

    [Column("medical_aid_deduction")]
    public double MedicalAidDeduction { get; set; }

    [Column("pension_deduction")]
    public double PensionDeduction { get; set; }

    [Column("union_deduction")]
    public double UnionDeduction { get; set; }

    [Column("pay_full_monthly_salary")]
    public bool PayFullMonthlySalary { get; set; }

    [Column("daily_hours")]
    public double DailyHours { get; set; } = 8.0;

    [Column("work_days_weekly")]
    public int WorkDaysWeekly { get; set; } = 5;

    [Column("email")]
    public string? Email { get; set; }

    [Column("phone")]
    public string? Phone { get; set; }

    [Column("manager_user_id")]
    public Guid? ManagerUserId { get; set; }

    [Column("company_id")]
    public Guid CompanyId { get; set; }

    [Column("user_id")]
    public Guid? UserId { get; set; }

    [Column("is_active")]
    public bool IsActive { get; set; } = true;

    [Column("profile_photo_url")]
    public string? ProfilePhotoUrl { get; set; }

    [Column("id_number")]
    public string? IdNumber { get; set; }

    [Column("bank_account")]
    public string? BankAccount { get; set; }

    [Column("bank_name")]
    public string? BankName { get; set; }

    [Column("bank_branch_code")]
    public string? BankBranchCode { get; set; }

    [Column("bank_details_updated_at")]
    public DateTime? BankDetailsUpdatedAt { get; set; }

    [Column("bank_details_updated_by")]
    public string? BankDetailsUpdatedBy { get; set; }

    [Column("shift_template_id")]
    public Guid? ShiftTemplateId { get; set; }

    [Column("login_password_ready")]
    public bool LoginPasswordReady { get; set; }

    [Column("registration_status")]
    [JsonProperty("registration_status")]
    public string RegistrationStatus { get; set; } = "active";

    [Column("created_at")]
    public DateTime CreatedAt { get; set; }

    // Computed — [JsonIgnore] prevents these from being serialized into INSERT/UPDATE bodies
    [JsonIgnore]
    public EmploymentType EmploymentType => EmploymentTypeRaw switch
    {
        "part-time" or "partTime" => EmploymentType.PartTime,
        "contract" => EmploymentType.Contract,
        "student" => EmploymentType.Student,
        _ => EmploymentType.Permanent
    };

    [JsonIgnore]
    public AccessLevel AccessLevel => AccessLevelRaw switch
    {
        "manager" => AccessLevel.Manager,
        "admin" => AccessLevel.Admin,
        "hr_admin" or "hrAdmin" => AccessLevel.HrAdmin,
        "owner" => AccessLevel.Owner,
        _ => AccessLevel.Employee
    };

    [JsonIgnore]
    public WorkerType WorkerType => WorkerTypeRaw switch
    {
        "contractor" => WorkerType.Contractor,
        "subcontractor" => WorkerType.Subcontractor,
        _ => WorkerType.Employee
    };

    [JsonIgnore]
    public string FullName => $"{Name} {Surname}".Trim();

    [JsonIgnore]
    public string Initials
    {
        get
        {
            var n = Name.Length > 0 ? Name[0].ToString() : "";
            var s = Surname.Length > 0 ? Surname[0].ToString() : "";
            return (n + s).ToUpperInvariant();
        }
    }

    [JsonIgnore]
    public string EmploymentTypeDisplay => EmploymentTypeRaw switch
    {
        "part-time" or "partTime" => "Part-Time",
        "contract" => "Contract",
        "student" => "Student",
        _ => "Permanent"
    };

    [JsonIgnore]
    public string AccessLevelDisplay => AccessLevelRaw switch
    {
        "owner" => "Owner",
        "hr_admin" or "hrAdmin" => "HR Admin",
        "admin" => "Admin",
        "manager" => "Manager",
        _ => "Employee"
    };

    [JsonIgnore]
    public string RequestedDateDisplay =>
        CreatedAt.Year < 2000 ? "—" : CreatedAt.ToLocalTime().ToString("dd MMM yyyy");

    [JsonIgnore]
    public bool IsHr => AccessLevel is AccessLevel.HrAdmin or AccessLevel.Owner or AccessLevel.Admin;

    /// <summary>Owner, HR, Admin, or field/ops Manager — uses company (HR/operations) dashboard.</summary>
    [JsonIgnore]
    public bool UsesCompanyDashboard =>
        AccessLevel is AccessLevel.Owner or AccessLevel.HrAdmin or AccessLevel.Admin or AccessLevel.Manager;

    [JsonIgnore]
    public bool IsFieldManager => AccessLevel == AccessLevel.Manager;

    [JsonIgnore]
    public bool IsManager => AccessLevel is AccessLevel.Manager or AccessLevel.HrAdmin or AccessLevel.Owner or AccessLevel.Admin;

    [JsonIgnore]
    public bool HasBankingDetails =>
        !string.IsNullOrWhiteSpace(BankAccount)
        || !string.IsNullOrWhiteSpace(BankName)
        || !string.IsNullOrWhiteSpace(BankBranchCode);

    [JsonIgnore]
    public string MaskedBankAccount => BankDetailsFormatting.MaskAccount(BankAccount);

    [JsonIgnore]
    public string BankDetailsUpdatedDisplay =>
        BankDetailsUpdatedAt.HasValue
            ? $"Updated {BankDetailsUpdatedAt.Value.ToLocalTime():dd MMM yyyy HH:mm} by {BankDetailsFormatting.UpdatedByLabel(BankDetailsUpdatedBy)}"
            : "";

    [JsonIgnore]
    public bool HasBankDetailsAudit => BankDetailsUpdatedAt.HasValue;

    /// <summary>Import preview only — not persisted.</summary>
    [JsonIgnore]
    public string? ImportTimeTemplateName { get; set; }
}
